backend/tournesol/suggestions/suggester.py

```diff
@@ -0,0 +1,236 @@
+        comparison graph
+        """
+        recommendation_user = RecommendationUser(
+            self._entity_to_video, new_user, self.criteria, self.poll
+        )
+        self._user_specific_graphs[new_user] = Graph(
+            local_user=recommendation_user,
+            local_poll=self.poll,
+            local_criteria=self.criteria,
+        )
+        # build user graph
+        query: QuerySet = (
+            ComparisonCriteriaScore.objects
+                                   .filter(comparison__poll__name=self.poll.name)
+                                   .filter(criteria=self.criteria)
+                                   .filter(comparison__user__email=new_user.email)
+        )
+        for c in query:
+            va = self._entity_to_video[c.comparison.entity_1]
+            vb = self._entity_to_video[c.comparison.entity_2]
+            self._user_specific_graphs[new_user].add_edge(va, vb)
+
+    def register_user_comparison(self, user: User, va: SuggestedVideo, vb: SuggestedVideo):
+        """
+        Function used to register a comparison submitted by the user, to keep the complete graph
+        up to date
+        """
+        self._complete_graph.add_edge(va, vb)
+        if user in self._user_specific_graphs.keys():
+            self._user_specific_graphs[user].add_edge(va, vb)
+
+    def get_first_video_recommendation(
+            self,
+            user: User,
+            nb_video_required: int
+    ) -> list[SuggestedVideo]:
+        """
+        Function used to get the first video recommendation for the requested user and returning
+        nb_video_required videos
+        """
+        # Lazily load the user graph
+        if user not in self._user_specific_graphs.keys():
+            self.register_new_user(user)
+        result = []
+
+        # Give the first video id to the graph so the sorting will take that into account
+        user_graph = self._user_specific_graphs[user]
+        user_graph.compute_offline_parameters(self._get_user_comparability_augmenting_videos())
+        user_graph.prepare_for_sorting()
+
+        # Prepare the set of videos to sort, taking the videos present in the graph
+        # and append the ones that are not yet compared by the user
+        considered_vids_list = self._prepare_video_list(user)
+
+        max_vid_pref = 0
+        # Todo : take into account the rate later list / already seen videos ?
+        for v in considered_vids_list:
+            v.user_pref = max(v.nb_comparison_with.values()) / v.comparison_nb
+            if v.user_pref > max_vid_pref:
+                max_vid_pref = v.user_pref
+
+        for i in range(nb_video_required):
+            for v in considered_vids_list:
+                act_user_pref = v.user_pref
+                if (
+                        act_user_pref
+                        >= (nb_video_required - i) / (nb_video_required + 1) * max_vid_pref
+                        and v not in result
+                ):
```

Comment:
I don't understand what's going on here?

Reply:
I refactored this part, let me know if it is clearer this way

Reply:
Ok, but what is still unclear is why we need to define this arbitrary threshold when including the videos in the results. What's the motivation for this instead of simply using the top results ranking by highest "user_pref"?